server/api/utils/query-builder.spec.js

```diff
@@ -0,0 +1,135 @@
+    });
+
+    it('should check that the provided filter function returns a result', async function() {
+
+      const valueFunc = stub().returns(42);
+      const filterFunc = stub(); // Will be called but returns no result.
+      const filter = filterFactory(valueFunc, filterFunc);
+
+      const mockQuery = {};
+      const mockContext = {
+        get: stub().returns(mockQuery),
+        set: stub()
+      };
+
+      await expect(filter(mockContext)).to.eventually.be.rejectedWith('The filter function returned a falsy result');
+
+      expect(valueFunc).to.have.been.calledWithExactly(mockContext);
+      expect(valueFunc).to.have.callCount(1);
+      expect(mockContext.get).to.have.been.calledWithExactly('query');
+      expect(mockContext.get).to.have.been.calledImmediatelyAfter(valueFunc);
+      expect(mockContext.get).to.have.callCount(1);
+      expect(filterFunc).to.have.been.calledWithExactly(mockQuery, 42, mockContext);
+      expect(filterFunc).to.have.been.calledImmediatelyAfter(mockContext.get);
+      expect(filterFunc).to.have.callCount(1);
+      expect(mockContext.set).to.have.callCount(0); // Make sure the query was not modified.
+    });
+
+    it('should check that the provided filter function returns a valid query', async function() {
+
+      const valueFunc = stub().returns(42);
+      const filterFunc = stub().returns('foo'); // Returns something that is not a query.
+      const filter = filterFactory(valueFunc, filterFunc);
+
+      const mockQuery = {};
+      const mockContext = {
+        get: stub().returns(mockQuery),
+        set: stub()
+      };
+
+      await expect(filter(mockContext)).to.eventually.be.rejectedWith('The modified query is not of the same type');
+
+      expect(valueFunc).to.have.been.calledWithExactly(mockContext);
+      expect(valueFunc).to.have.callCount(1);
+      expect(mockContext.get).to.have.been.calledWithExactly('query');
+      expect(mockContext.get).to.have.been.calledImmediatelyAfter(valueFunc);
+      expect(mockContext.get).to.have.callCount(1);
+      expect(filterFunc).to.have.been.calledWithExactly(mockQuery, 42, mockContext);
+      expect(filterFunc).to.have.been.calledImmediatelyAfter(mockContext.get);
+      expect(filterFunc).to.have.callCount(1);
+      expect(mockContext.set).to.have.callCount(0); // Make sure the query was not modified.
+    });
+
+    it('should check that the provided filter function returns a query of the same type', async function() {
+
+      // Prepare two different models.
+      const FirstModel = Abstract.extend({});
+      const SecondModel = Abstract.extend({});
+
+      const valueFunc = stub().returns(42);
+      const filterFunc = stub().returns(new SecondModel()); // Return a query that is not of the same model as the base query.
+      const filter = filterFactory(valueFunc, filterFunc);
+
+      const mockQuery = new FirstModel(); // The base query.
```

Comment:
I still have a hard time with the fact that instanciating a Model equals creating a new query.